src/ws-api.ts

```diff
@@ -35,127 +30,14 @@ export class WsApi {
-          return socket.send(responseBuffer);
-        }
-
-        dwnRequest = JSON.parse(dwnRequest);
-      } catch (e) {
-        const jsonRpcResponse = createJsonRpcErrorResponse(
-          uuidv4(),
-          JsonRpcErrorCodes.BadRequest,
-          e.message,
-        );
-
-        const responseBuffer = WsApi.jsonRpcResponseToBuffer(jsonRpcResponse);
-        return socket.send(responseBuffer);
-      }
-
-      // Check whether data was provided in the request
-      const { encodedData } = dwnRequest.params;
-      const requestDataStream = encodedData
-        ? DataStream.fromBytes(base64url.baseDecode(encodedData))
-        : undefined;
-
-      const requestContext: RequestContext = {
-        dwn,
-        transport: 'ws',
-        dataStream: requestDataStream,
-      };
-      const { jsonRpcResponse } = await jsonRpcApi.handle(
-        dwnRequest,
-        requestContext,
-      );
-
-      if (jsonRpcResponse.error) {
-        requestCounter.inc({ method: dwnRequest.method, error: 1 });
-      } else {
-        requestCounter.inc({
-          method: dwnRequest.method,
-          status: jsonRpcResponse?.result?.reply?.status?.code || 0,
-        });
-      }
-
-      const responseBuffer = WsApi.jsonRpcResponseToBuffer(jsonRpcResponse);
-      return socket.send(responseBuffer);
-    });
-  }
-
-  /**
-   * This handler returns an interval to ping clients' socket every 30s
-   * if a pong hasn't received from a socket by the next ping, the server will terminate the socket connection.
-   */
-  #setupHeartbeat(): NodeJS.Timer {
-    // Sometimes connections between client <-> server can get borked in such a way that
-    // leaves both unaware of the borkage. ping messages can be used as a means to verify
-    // that the remote endpoint is still responsive. Server will ping each socket every 30s
-    // if a pong hasn't received from a socket by the next ping, the server will terminate
-    // the socket connection
-    return setInterval(() => {
-      this.#wsServer.clients.forEach(function (socket) {
-        if (socket[SOCKET_ISALIVE_SYMBOL] === false) {
-          return socket.terminate();
-        }
-
-        socket[SOCKET_ISALIVE_SYMBOL] = false;
-        socket.ping();
-      });
-    }, HEARTBEAT_INTERVAL);
-  }
-
   /**
    * Handler for starting a WebSocket.
    * Sets listeners for `connection`, `close` events.
    * It clears `heartbeatInterval` when a `close` event is made.
    */
   #setupWebSocket(): void {
-    this.#wsServer.on('connection', this.#handleConnection.bind(this));
-
-    const heartbeatInterval = this.#setupHeartbeat();
-
-    this.#wsServer.on('close', function close() {
-      clearInterval(heartbeatInterval);
-    });
+    this.#wsServer.on('connection', this.#connections.connect.bind(this));
+    this.#wsServer.on('close', this.#connections.close.bind(this));
   }
 
   start(callback?: () => void): WebSocketServer {
```

Comment:
Existing code, but while at it: I don't see the point of making this more complicate than it needs to be, there is nothing async in this method, consider removing the callback, if anything turn the method into async so caller can await if needed.

Reply:
I'm glad you said this. I was considering removing it, and think I did but re-added it to reduce review burden.
I noticed the same on the `DwnServer.start()`, although that function is async, I don't think the callback does anything special and can be removed also. Thoughts?

If we wanted to do a callback before some other things happened within the promise i'd understand, but isn't it the same as just adding a .then(callback) when you don't want to await the promise?

Reply:
Removed from here and `DwnServer.start()`